### PR TITLE
Add Validation of EncryptedShares for Dispute Resolution

### PR DESCRIPTION
**Pull request by tschudid**
## Purpose

The goal of this pull request is to add a public validation method for encrypted shares. This method can be used for dispute resolution between guardians in the key generation ceremony.

## Changes

- Add missing share validation to the share decryption function.
- Add public validation method for encrypted shares

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.

### CODE REVIEW

File: src/eg/src/guardian_share.rs [hunk 1]
```diff
@@ -226,14 +290,91 @@ impl GuardianEncryptedShare {
...
+        let vec_k_i_j = &dealer_public_key.coefficient_commitments.0;
+        let rhs = vec_k_i_j
+            .iter()
+            .enumerate()
+            .fold(BigUint::one(), |prod, (j, k_i_j)| {
+                //This is fine as j < k
+                #[allow(clippy::unwrap_used)]
+                let j: u32 = j.try_into().unwrap();
+                (prod * k_i_j.0.modpow(&l.pow(j), p)) % p
+            });
```
Comment (abizjak) [inline src/eg/src/guardian_share.rs:367]:
```suggestion
        let rhs = (0u32..).zip(&vec_k_i_j)
            .fold(BigUint::one(), |prod, (j, k_i_j)| {
                //This is fine as j < k
                (prod * k_i_j.0.modpow(&l.pow(j), p)) % p
            });
```

**Review by abizjak — APPROVED** [review]

File: src/eg/src/guardian_share.rs [hunk 2]
```diff
@@ -128,62 +155,99 @@ impl GuardianEncryptedShare {
         eg_hmac(&k0, &v)
     }
 
-    /// This function creates a new [`GuardianEncryptedShare`] of the dealer's secret key for a given recipient.
-    /// 
+    /// This function computes a [`GuardianEncryptedShare`] of a share for a given recipient and nonce.
+    ///
     /// The arguments are
```
Comment (hamiidreza) [inline src/eg/src/guardian_share.rs:160]: Arguments described in the documentation doesn't seem to be consistent with the code!

File: src/eg/src/guardian_share.rs [hunk 3]
```diff
@@ -128,62 +155,99 @@ impl GuardianEncryptedShare {
...
-        let x = &BigUint::from(l);
-        let mut p_l = BigUint::from(0_u8);
-        for coeff in dealer_private_key.secret_coefficients.0.iter().rev() {
-            p_l = (p_l * x + &coeff.0) % q;
-        }
-
         //Ciphertext as in Equation (19)
-        let c1 = xor(to_be_bytes_left_pad(&p_l, 32).as_slice(), k1.0.as_slice());
+        let c1 = xor(to_be_bytes_left_pad(share, 32).as_slice(), k1.0.as_slice());
         //The unwrap is justified as the output the XOR will always be 32 bytes.
```
Comment (hamiidreza) [inline src/eg/src/guardian_share.rs:192]:
```suggestion
        //The unwrap is justified as the output of the XOR will always be 32 bytes.
```

File: src/eg/src/guardian_share.rs [hunk 4]
```diff
@@ -226,14 +290,87 @@ impl GuardianEncryptedShare {
...
+                (prod * k_i_j.0.modpow(&l.pow(j), p)) % p
+            });
+        if g_p_l != rhs {
+            return Err(DecryptionError::InvalidShare);
+        }
+
+        Ok(p_l)
+    }
+
+    /// This function validates a [`GuardianEncryptedShare`] with respect to given [`GuardianEncryptionSecret`] and [`GuardianPublicKey`] of dealer and recipient in case of a dispute.
```
Comment (hamiidreza) [inline src/eg/src/guardian_share.rs:318]: Could we add more into the documentation about the output of the function? E.g., something along the lines of "This function returns false if the dealer's published secrets do not match the published ciphertext, or the equation `21`. If they match, the function returns true, meaning the recipient's claim about the wrongdoing by the dealer is dismissed."

File: src/eg/src/guardian_share.rs [hunk 5]
```diff
@@ -456,4 +608,56 @@ mod test {
 
         assert_eq!(Some(joint_key_1), joint_key_2, "Joint keys should match.")
     }
+
+    #[test]
+    fn test_public_validation() {
+        let mut csprng = Csprng::new(b"test_proof_generation");
```
Comment (hamiidreza) [inline src/eg/src/guardian_share.rs:614]:
```suggestion
        let mut csprng = Csprng::new(b"test_public_validation");
```

**Review by hamiidreza — APPROVED** [review]
LGTM

File: src/eg/src/guardian_share.rs (hunk 2 above)
Comment (tschudid) [reply src/eg/src/guardian_share.rs:160]: Updated the comments to match the code.

File: src/eg/src/guardian_share.rs (hunk 4 above)
Comment (tschudid) [reply src/eg/src/guardian_share.rs:317]: I have added your suggestion.

**Pull request merged**